pkg/distributor/distributor_test.go

```diff
@@ -4334,3 +4362,280 @@ func countMockIngestersCalls(ingesters []mockIngester, name string) int {
 	}
 	return count
 }
+
+// TestDistributor_MetricsWithRequestModifications tests that the distributor metrics are properly updated when
+// requests get modified by the mechanisms that can modify them: relabel rules, drop labels, ha-dedupe, forwarding, limits.
+func TestDistributor_MetricsWithRequestModifications(t *testing.T) {
+	tenant := "tenant1"
+	getCtx := func() context.Context {
+		return user.InjectOrgID(context.Background(), tenant)
+	}
+	getDefaultConfig := func() prepConfig {
+		var limits validation.Limits
+		flagext.DefaultValues(&limits)
+		limits.MaxGlobalExemplarsPerUser = 1000
+
+		return prepConfig{
+			limits:            &limits,
+			numIngesters:      1,
+			happyIngesters:    1,
+			replicationFactor: 1,
+			numDistributors:   1,
+		}
+	}
+	getDistributor := func(config prepConfig) (*Distributor, *prometheus.Registry) {
+		ds, _, regs := prepare(t, config)
+		return ds[0], regs[0]
+	}
+	getExpectedMetrics := func(requestsIn, samplesIn, exemplarsIn, metadataIn, receivedRequests, receivedSamples, receivedExemplars, receivedMetadata int) (string, []string) {
```

Comment:
I think this would be a little easier to understand if it took a single config struct as an argument. Then you could invoke it like

```go
expected = getExpectedMetrics(expectedCfg{
    requestsIn: 10,
    samplesIn: 10,
    /* ... */
})
```

Reply:
good idea, that would make it more obvious what each parameter is, I'll update that.

Reply:
done here 5e8a0391d500c57b47e749e0aaf1a6dff242ace7